dotnetv3/EC2/VirtualPrivateCloudExamples/CreateVPCforS3Example/CreateVPCforS3.cs

```diff
@@ -0,0 +1,136 @@
+    /// </summary>
+    /// <param name="configuration">Configuration to specify resource ids.</param>
+    /// <param name="ec2Client">Initialized EC2 client.</param>
+    /// <returns>The S3 url using the endpoint.</returns>
+    public static async Task<string?> CreateVPCforS3Client(IConfiguration configuration,
+        AmazonEC2Client ec2Client)
+    {
+        var endpointResponse = await ec2Client.CreateVpcEndpointAsync(
+            new CreateVpcEndpointRequest
+            {
+                VpcId = configuration["VpcId"],
+                VpcEndpointType = VpcEndpointType.Interface,
+                ServiceName = "com.amazonaws.us-east-1.s3",
+                SubnetIds = new List<string> { configuration["SubnetId"]! },
+                SecurityGroupIds = new List<string> { configuration["SecurityGroupId"]! },
+                TagSpecifications = new List<TagSpecification>
+                {
+                    new TagSpecification
+                    {
+                        ResourceType = ResourceType.VpcEndpoint,
+                        Tags = new List<Tag>
+                        {
+                            new Tag("service", "S3")
+                        }
+                    }
+                }
+            });
+
+        var newEndpoint = endpointResponse.VpcEndpoint;
+
+        Console.WriteLine(
+            $"VPC Endpoint {newEndpoint.VpcEndpointId} was created, waiting for it to be available. " +
+            $"This may take a few minutes.");
+
+        State? endpointState = null;
+
+        while (endpointState == null ||
+               !(endpointState == State.Available || endpointState == State.Failed))
+        {
+            var endpointDescription = await ec2Client.DescribeVpcEndpointsAsync(
+                new DescribeVpcEndpointsRequest
+                {
+                    VpcEndpointIds = new List<string>
+                        { endpointResponse.VpcEndpoint.VpcEndpointId }
+                });
+            endpointState = endpointDescription.VpcEndpoints.FirstOrDefault()?.State;
+            Thread.Sleep(500);
+        }
+
+        if (endpointState == State.Failed)
+        {
+            Console.WriteLine("VPC Endpoint failed.");
+            return null;
+        }
+
+        // For newly created endpoints we may need to wait a few
+        // more minutes before using it for the Amazon S3 client.
+        Thread.Sleep(int.Parse(configuration["WaitTimeMilliseconds"]!));
+
+        // Return the endpoint to create a ServiceURL to use with the S3 client.
+        var vpceUrl =
+            @$"https://bucket{endpointResponse.VpcEndpoint.DnsEntries[0].DnsName.Trim('*')}/";
+
+        return vpceUrl;
+    }
+
+    /// <summary>
+    /// Create an Amazon S3 client with a VPC url.
+    /// This code must be executed in an environment that has access to the VPC in order to successfully
+    /// list the objects in the Amazon S3 bucket when using a private VPC endpoint.
```

Comment:
/// To successfully list the objects in the S3 bucket when you're using a private VPC endpoint,
    /// you must run this code in an environment that has access to the VPC.

Reply:
Fixed.